# Programming Lesson and Exercises - ORG

The goal of these programming lesson and exercises is to teach you how to organise and query data in a relational database using Python and SQL.


This notebook has the following structure:

- The first part introduces the concepts for this week. The theory is interleaved with small exercises, which have the goal of letting you practice the concepts that were just intruduced.
- At the end there are one or more larger exercises, which have the goal to test what you have learned earlier. These will be more difficult and will require more independent work than the exercises in the first part.

All exercises can be solved with the concepts that were introduced earlier. Since there are often more than one correct way to solve a programming problem, we try to accept various correct anwers. However, many of the automatic tests in Momotor (see _How to submit your work_ below) assume that your answers are constructed using the concepts introduced in these notebooks. If you look for answers on the Internet (e.g. if you import other libraries) you run the risk that your answers will be rejected.

Some of the small exercises can be solved by copy-pasting code from the examples. However, it is up to you to try to solve the exercises yourself, which will help you learn, before copy-pasting the answers. The ease of looking up answers is meant to provide guidance when you get stuck, especially for those of you who are new to programming.

For your convenience, in the `support` directory you will find a summary of the Python methods introduced in this notebook.


<div class="alert-info alert" role="alert-info">Make sure to consult the lecture slides for more details on the relational database model and the SQL langauge.</div>


# Introduction to This Template Notebook

* This is a **personal** notebook.
* Make sure you work in a **copy** of `...-template.ipynb`,
**renamed** to `...-yourIDnr.ipynb`,
where `yourIDnr` is your TU/e identification number.

<div class="alert alert-danger" role="danger">
<h3>Integrity</h3>
<ul>
    <li>In this course you must act according to the rules of the TU/e code of scientific conduct.</li>
    <li>All the exercises and the graded assignments are to be executed individually and independently.</li>
    <li>You must not copy from the Internet, your friends, books... If you represent other people's work as your own, then that constitutes fraud and will be reported to the Examination Committee.</li>
    <li>Making your work available to others (complicity) also constitutes fraud.</li>
</ul>
</div>

You are expected to work with Python code in this notebook.

The locations where you should write your solutions can be recognized by
**marker lines**,
which look like this:

>`#//`
>    `BEGIN_TODO [Label]` `Description` `(n points)`
>
>`#//`
>    `END_TODO [Label]`

<div class="alert alert-warning" role="alert">Do NOT modify or delete these marker lines.  Keep them as they are.<br/>
NEVER write code <i>outside</i> the marked blocks.
Such code cannot be evaluated.
</div>

Proceed in this notebook as follows:
* **Read** the text.
* **Fill in** your solutions between `BEGIN_TODO` and `END_TODO` marker lines.
* **Run** _all_ code cells (also the ones _without_ your code),
    _in linear order_ from the first code cell.

**Personalize your notebook**:
1. Copy the following three lines of code:

  ```python
  AUTHOR_NAME = 'Your Full Name'
  AUTHOR_ID_NR = '1234567'
  AUTHOR_DATE = 'YYYY-MM-DD'
  ```
1. Paste them between the marker lines in the next code cell.
1. Fill in your _full name_, _identification number_, and the current _date_ (i.e. when you first modified this notebook, e.g. '2023-11-01') as strings between the `Author` markers.
1. Run the code cell by putting the cursor there and typing **Control-Enter**.


In [1]:
#// BEGIN_TODO [Author] Name, Id.nr., Date, as strings (1 point)

AUTHOR_NAME = 'Shiyun Kong'
AUTHOR_ID_NR = '1920847'
AUTHOR_DATE = '2023-12-07'
#// END_TODO [Author]

## Table of Contents

- [Learning Objectives](#Learning-Objectives)
- [1. SQL Basics](#1.-SQL-Basics)
    - [Introduction](#Introduction)
    - [Loading the Libraries](#Loading-the-Libraries)
    - [Common Steps in Working with SQLite Databases](#Common-Steps-in-Working-with-SQLite-Databases)
    - [Step 1. Connect to Database](#Step-1.-Connect-to-Database)
    - [Step 2. Operate on Database](#Step-2.-Operate-on-Database)
        - [`SELECT` Statement](#SELECT-Statement)
        - [`WHERE` Clause](#WHERE-Clause)
        - [Exercise 1.a](#Exercise-1.a)
        - [Exercise 1.b](#Exercise-1.b)
        - [`GROUP BY` Clause](#GROUP-BY-Clause)
        - [Exercise 1.c](#Exercise-1.c)
        - [`ORDER BY` Clause](#ORDER-BY-Clause)
        - [Exercise-1.d](#Exercise-1.d)
        - [`LIMIT` Clause](#LIMIT-Clause)
        - [Exercise 1.e](#Exercise-1.e)
        - [Joining Tables](#Joining-Tables)
        - [Exercise 1.f](#Exercise-1.f)
        - [Exercise 1.g](#Exercise-1.g)
        - [Exercise 1.h](#Exercise-1.h)
        - [Subqueries or Nested Queries](#Subqueries-or-Nested-Queries)
        - [Exercise 1.i](#Exercise-1.i)
    - [Step 3. Close Connection to Database](#Step-3.-Close-Connection-to-Database)
- [2. Data Cleaning](#2.-Data-Cleaning)
    - [`df.astype()`: Convert to Given Type](#df.astype():-Convert-to-Given-Type)
    - [Exercise 2.a](#Exercise-2.a)
    - [`pd.to_numeric()`: Convert to Numeric](#pd.to_numeric():-Convert-to-Numeric)
    - [`ts.unique()`: Find All Unique Values](#ts.unique():-Find-All-Unique-Values)
    - [Exercise 2.b](#Exercise-2.b)
    - [`df.duplicated()`: Find Duplicates](#df.duplicated():-Find-Duplicates)
    - [Exercise 2.c](#Exercise-2.c)
    - [`df.drop_duplicates()`: Remove Duplicates](#df.drop_duplicates():-Remove-Duplicates)
    - [Exercise 2.d](#Exercise-2.d)
    - [`df.dropna()`: Remove Rows with Missing Values](#df.dropna():-Remove-Rows-with-Missing-Values)
    - [Exercise 2.e](#Exercise-2.e)
    - [`df.fillna()`: Replace Missing Values](#df.fillna():-Replace-Missing-Values)
    - [Exercise 2.f](#Exercise-2.f)
    - [`ts.str.strip()`: Strip Characters from Begin/End of Strings](#ts.str.strip():-Strip-Characters-from-Begin/End-of-Strings)
    - [Exercise 2.g](#Exercise-2.g)
- [3. Exercises: Cleaning the Mouse Experiment](#3.-Exercises:-Cleaning-the-Mouse-Experiment)
    - [Exercise 3.a](#Exercise-3.a)
    - [Exercise 3.b](#Exercise-3.b)
    - [Exercise 3.c](#Exercise-3.c)
    - [Exercise 3.d](#Exercise-3.d)
    - [Exercise 3.e](#Exercise-3.e)
    - [Exercise 3.f](#Exercise-3.f)
    - [Exercise 3.g](#Exercise-3.g)
- [4. Exercises: Analyzing the Mouse Experiment](#4.-Exercises:-Analyzing-the-Mouse-Experiment)
    - [Exercise 4.a](#Exercise-4.a)
    - [Exercise 4.b](#Exercise-4.b)
    - [Exercise 4.c](#Exercise-4.c)
    - [Exercise 4.d](#Exercise-4.d)
    - [Exercise 4.e](#Exercise-4.e)
    - [Exercise 4.f](#Exercise-4.f)
    - [Exercise 4.g](#Exercise-4.g)
    - [Exercise 4.h](#Exercise-4.h)

## Learning Objectives

After this lesson, you should 

* know how to work with SQLite databases through Pandas data frames;
* know how to query data in an SQLite database using Python and its `sqlite3` library;
* know how to perform basic data cleaning tasks using Pandas.

## 1. SQL Basics

### Introduction

So far, we have dealt with data contained in a _single table_,
typically stored in a CSV file and operated on in a Pandas `DataFrame`.
Often, however, the use of a single table leads to unnecessary duplication of data
and difficulties in data processing.

For example,
consider an experiment where we collect data from various patients.
We have data pertaining to each patient:
name, date of birth, and daily weight and temperature measurements.
How to _organize_ that data and store it?
This is important for the _data collection_ process and subsequent
_data analysis_ process.

You could try to put it in one table,
with one row per patient, and columns for name, date of birth, and each (day, measurement)-pair:

| Name | Date of Birth | Day 1 Weight | Day 1 Temperature | Day 2 Weight | Day 2 Temperature | ... |
| :--- | :------------ | -----------: | ----------------: | -----------: | ----------------: | --: |
| A | 1950-01-01 | 60.0 | 37.8 | 59.8 | 38.1 | ... |
| B | 1947-07-15 | 84.3 | 39.2 | 85.0 | 38.5 | ... |
| ... | ... | ... | ... | ... | ... | ... |

The number of columns would grow, and could differ per patient.

You could try to put it in one table,
with one row per patient per day,
and columns for name, date of birth, day, weight, and temperature,
repeating values that stay the same (name and date of birth):

| Name | Date of Birth | Day | Weight | Temperature |
| :--- | :------------ | --: | -----: | ----------: |
| A | 1950-01-01 | 1 | 60.0 | 37.8 |
| B | 1947-07-15 | 1 | 84.3 | 39.2 |
| A | 1950-01-01 | 2 | 59.8 | 38.1 |
| B | 1947-07-15 | 2 | 84.3 | 39.2 |
| ... | ... | ... | ... | ... |

In this approach, some data (necessarily) gets duplicated.

Also, note that patients may not be uniquely identified by name and date of birth.
In the first table, their measurements would still be kept apart.
But in the second table, this starts to be problematic.

Using multiple, related tables is a better approach.
In this case, two tables could be used:

**Patient data**

| ID | Name | Date of Birth |
| -: | :--- | :------------ |
| 13 | A | 1950-01-01 |
| 72 | B | 1947-07-15 |
| ... | ... | ... |

**Measurement data**

| ID | Day | Weight | Temperature |
| -: | --: | -----: | ----------: |
| 13 |  1 | 60.0 | 37.8 |
| 72 |  1 | 84.3 | 39.2 |
| 13 |  2 | 59.8 | 38.1 |
| 72 |  2 | 84.3 | 39.2 |
| ... | ... | ... | ... |

The additional column `ID` relates the two tables,
and helps to keep apart patients with the same name and date of birth.
It functions as the *key* of these two tables.

We assume that you have a basic understanding of how a _relational database_ is organized:

* a collection of (named, related) *tables*, where
* each table consists of a collection of *rows*, where
* each row contains one _value_ per *column*,
* where all values in a column belong to a prescribed *type*.

In this lesson we will use **SQL**, which stands for **Structured Query Language**,
a standardized language to work with _relational databases_.

We will use **SQLite** as _SQL database engine_
(it knows SQL, though not all of it).
Note that SQLite is a separate piece of open-source software,
that is installed as part of Anaconda.

### Loading the Libraries

To show examples, we load some Data Analytics libraries first:

In [2]:
import numpy as np
import pandas as pd

# Reveal a hint only while holding the mouse down
from IPython.display import HTML
HTML("<style>.h,.c{display:none}.t{color:#296eaa}.t:active+.h{display:block;}</style>")

Next, we load **`sqlite3`**, a Python library to work with _SQLite_ databases:

In [3]:
import sqlite3

### Common Steps in Working with SQLite Databases

Typically, the following sequence of steps takes place when working with an SQL database:

1. Connect to the database (residing in a file).
2. Operate on the database:
    * execute queries, to retrieve selected values;
    * create, update, or delete rows/values in a table;
    * create, alter (by adding or deleting columns), or delete tables.
    
3. Close the connection to the database.

Step 2 can be repeated as desired.

## Step 1. Connect to Database

In order to work with a database, a _connection_ to that database needs to be established.

In these exercises we will work with the SQLite database engine. An SQLite database is stored in a (binary) file in the file system. You connect to it via

* **`sqlite3.connect(file_name)`**

This function returns a **connection object**, that will be used in subsequent operations.

Here we open a connection to the database stored in the file `'countries.db'`, and assign this connection to a variable named `conn_countries`:

In [4]:
conn_countries = sqlite3.connect('datasets/countries.db')

> **Note:** If the database file does not exist, it will be created, starting out with an empty database.

The `'countries.db'` database contains two tables: `'continents'` and `'countries'`. The `'continents'` table has following columns:

- `'code'`: A two letter code of the continent.
- `'name'`: The name of the continent.

The `'countries'` table has the following columns:

- `'name'`: The name of the country.
- `'alpha_3'`: The ISO 3166-1 alpha-3 code of the country.
- `'tld'`: The top-level domain of the country.
- `'continent'`: A two letter code of the continent where the country is located (corresponding to the codes in the `'continents'` table).
- `'capital'`: The name of the capital of the country.
- `'area'`: The area of the country in square kilometers.
- `'population'`: The population size of the country.

## Step 2. Operate on Database

When working with an SQL database, operations are expressed as **SQL statements**. SQL defines many statements for organising and manipulating data in a database. In these exercises we will focus on querying data using the **SELECT** statement. See [W3 Schools - SQL Tutorial](https://www.w3schools.com/sql/) for more details about SQL.

General rules:

* SQL is _not_ case sensitive (`SELECT` and `select` are treated the same way).
* SQL statements are terminated by a semicolon ('**`;`**'). Note that many modern libraries, including Pandas, allow to skip the semicolon. However, it is a good practice to include it for portability.
* SQL identifiers that contain spaces, or that are the same as a keyword, are enclosed in **double quotes**.
* SQL string literals are enclosed in **single quotes**.

### `SELECT` Statement

The SQL **SELECT** statement retrieves values from a database and presents the result as a table organized in rows and columns.  
This statement has a number of required and optional clauses:

| Clause | Type | Remarks |
| :----- | :--: | :------ |
| `SELECT` _`expr1, expr2, ...`_ | Required | `SELECT *` selects all columns; `SELECT DISTINCT ...`. |
| `FROM` _`table1, table2, ...`_ | Required | |
| `WHERE` _`condition`_ | Optional | Filter on boolean condition before grouping. |
| `GROUP BY` _`column1, column2, ...`_ | Optional | |
| `ORDER BY` _`column1, column2, ...`_ | Optional | Append `DESC` for descending order. |
| `LIMIT` _`number`_ | Optional | Report only first given number of rows. |

Numeric expressions _`expr1`_ etc. can contain

* _column names_, which must be disambiguated when a column name appears in more than one table by prepending _table name_: `table1.column1`;
* aliases for renaming column/table names: e.g., `table1.column1 AS 'some name'` to rename `table1.column1` to `'some name'` (aliases exist only for the duration of the query);
* operators: `+`, `-`, `*`, `/`, `%`;
* functions: `MIN`, `MAX`, `COUNT`, `SUM`, `AVG` (these aggregate when `GROUP BY` is used).

An SQL query is stored in a Python string. Here is an example of a query that returns all the data in the `'countries'` table.

In [5]:
query_all = 'SELECT * FROM countries;'

The Pandas library offers a convenient function to work with SQL databases:

* **`pd.read_sql_query(query, connection)`**: execute an SQL `query` on an open `connection` to an SQL database, and return the query result as a `DataFrame`.

Here is an example of executing the above query `query_all` on the previously opened connection `conn_countries`:

In [6]:
pd.read_sql_query(query_all, conn_countries)

name alpha_3  tld continent           capital       area  \
0                 Andorra     AND  .ad        EU  Andorra la Vella      468.0   
1    United Arab Emirates     ARE  .ae        AS         Abu Dhabi    82880.0   
2             Afghanistan     AFG  .af        AS             Kabul   647500.0   
3     Antigua and Barbuda     ATG  .ag        NM        St. John's      443.0   
4                Anguilla     AIA  .ai        NM        The Valley      102.0   
..                    ...     ...  ...       ...               ...        ...   
246               Mayotte     MYT  .yt        AF         Mamoudzou      374.0   
247          South Africa     ZAF  .za        AF          Pretoria  1219912.0   
248                Zambia     ZMB  .zm        AF            Lusaka   752614.0   
249              Zimbabwe     ZWE  .zw        AF            Harare   390580.0   
250  Netherlands Antilles     ANT  .an        NM        Willemstad      960.0   

     population  
0         84000  
1       4975593  
2      29121286  
3         86754  
4         13254  
..          ...  
246      159042  
247    49000000  
248    13460305  
249    13061000  
250      300000  

[251 rows x 7 columns]

### `WHERE` Clause

The `WHERE` clause is a boolean condition, which can contain:

* _column names_;
* comparison operators: `=`, `<>`, `<`, `>`, `<=`, `>=`;
* boolean operators: `NOT`, `AND`, `OR`;
* pattern matching operator (to match string values against a pattern with wildcards): `LIKE`.
    * Example: the condition `... WHERE name LIKE "Nether%"` will select the rows that contain a string starting with `Nether` in the `name` column.

Here is an example of a query that selects all rows from the table `countries`, that have an area less than 1000 and a population over 200000, showing the columns `'name'` and `'capital'`:

In [7]:
query_area_population = '''
    SELECT name, capital
    FROM countries
    WHERE area < 1000 AND population > 200000;
'''
pd.read_sql_query(query_area_population, conn_countries)

name     capital
0              Barbados  Bridgetown
1               Bahrain      Manama
2                 Macao       Macao
3                 Malta    Valletta
4              Maldives        Male
5             Singapore   Singapore
6  Netherlands Antilles  Willemstad

> **Note:** The SQL query `query_area_population` is created as a *multiline string*, which is started and ended by 3 single or double quotes. This allows us to format text to maximize readability. The line breaks and indentation are included in the string, as can be seen by executing the cell below, but this extra whitespace is ignored by the parser.

In [8]:
query_area_population

'\n    SELECT name, capital\n    FROM countries\n    WHERE area < 1000 AND population > 200000;\n'

### Exercise 1.a

Execute the following SQL query to select the name and area of all countries with an area larger than one million square km:

```
SELECT name, area
FROM countries
WHERE area > 1000000
```

Execute the query on the open database connection `conn_countries`. Assign the resulting data frame to the variable `result_select`.

In [9]:
#// BEGIN_TODO [ORG_1a] Execute pre-defined query (1 point)

query = '''
    SELECT name, area
    FROM countries
    WHERE area > 1e6
'''
result_select = pd.read_sql_query(query, conn_countries)
#// END_TODO [ORG_1a]

result_select.head(10)

name        area
0                            Angola   1246700.0
1                        Antarctica  14000000.0
2                         Argentina   2766890.0
3                         Australia   7686850.0
4                           Bolivia   1098580.0
5                            Brazil   8511965.0
6                            Canada   9984670.0
7  Democratic Republic of the Congo   2345410.0
8                             China   9596960.0
9                          Colombia   1138910.0

### Exercise 1.b

Execute a query to select the name and population of European countries with a population larger than 60 million. Assign the query to the variable `query_where` and the result to the variable `result_where`.

In [10]:
#// BEGIN_TODO [ORG_1b] Define and execute a query (1 point)

query_where = '''
    SELECT name, population
    FROM countries
    WHERE population > 60e6 AND continent == "EU"
'''
result_where = pd.read_sql_query(query_where, conn_countries)
#// END_TODO [ORG_1b]

result_where

name  population
0         Germany    81802257
1          France    64768389
2  United Kingdom    62348447
3           Italy    60340328
4          Russia   140702000

### `GROUP BY` Clause

The `GROUP BY` clause aggregates data, similarly to the `.groupby()` method in Pandas. The `MIN`, `MAX`, `COUNT`, `SUM`, `AVG` functions are used in the `SELECT` expression to specify how to aggregate the data in the columns.

Here is an example query using the `GROUP BY` clause to obtain an overview of all continents, with for each continent:

* the continent code;
* the number of countries on it;
* the total population;
* the maximum population of its countries.

In [11]:
query = '''
    SELECT continent, COUNT(name), SUM(population), MAX(population)
    FROM countries
    GROUP BY continent
'''
pd.read_sql_query(query, conn_countries)

continent  COUNT(name)  SUM(population)  MAX(population)
0        AF           58       1016879730        154000000
1        AN            5              170              140
2        AS           52       4130586341       1330044000
3        EU           53        739141025        140702000
4        NM           42        540105043        310232863
5        OC           27         36066049         21515754
6        SM           14        400143568        201103330

### Exercise 1.c

Execute a query to obtain an overview of the continents, showing for each continent the continent code and the total area. Assign the query to the variable `query_group` and the result to the variable `result_group`.

In [12]:
#// BEGIN_TODO [ORG_1c] Define and execute a query (1 point)

query_group = '''
    SELECT continent, SUM(area)
    FROM countries
    GROUP BY continent
'''
result_group = pd.read_sql_query(query_group, conn_countries)
#// END_TODO [ORG_1c]

result_group

continent    SUM(area)
0        AF  30360296.00
1        AN  14012193.00
2        AS  31603228.70
3        EU  23079944.09
4        NM  24512514.00
5        OC   8523505.90
6        SM  17818508.00

### `ORDER BY` Clause

The `ORDER BY` clause orders the resulting rows in either ascending (`ASC`) or descending (`DESC`) order. The `ASC` or `DESC` argument can be skipped, in which case the default is ascending order.

Here is an example query using the `ORDER BY` clause to obtain an overview of countries sorted in descending order of their population size:

In [13]:
query = '''
    SELECT name, population
    FROM countries
    ORDER BY population DESC;
'''
pd.read_sql_query(query, conn_countries)

name  population
0                                           China  1330044000
1                                           India  1173108018
2                                   United States   310232863
3                                       Indonesia   242968342
4                                          Brazil   201103330
..                                            ...         ...
246  South Georgia and the South Sandwich Islands          30
247                                    Antarctica           0
248                                 Bouvet Island           0
249             Heard Island and McDonald Islands           0
250          United States Minor Outlying Islands           0

[251 rows x 2 columns]

### Exercise 1.d

Execute a query to select the name, continent code, population, and area of all
countries with an area larger than one million square km and a population less than ten million, ordered in ascending order of their area.
Assign the query to the variable `query_order` and the result to the variable `result_order`.

In [14]:
#// BEGIN_TODO [ORG_1d] Define and execute a query (1 point)

query_order = '''
    SELECT name, continent, population, area
    FROM countries
    WHERE area > 1e6 AND population < 10e6
    ORDER BY area ASC
'''
result_order = pd.read_sql_query(query_order, conn_countries)
#// END_TODO [ORG_1d]

result_order

name continent  population        area
0  Mauritania        AF     3205060   1030700.0
1     Bolivia        SM     9947418   1098580.0
2    Mongolia        AS     3086918   1565000.0
3       Libya        AF     6461454   1759540.0
4   Greenland        NM       56375   2166086.0
5  Antarctica        AN           0  14000000.0

### `LIMIT` Clause

The `LIMIT n` clause limits the result to only the first `n` rows. Here is an example query using the `LIMIT` clause to obtain an overview of top 10 countries sorted in descending order by population size:

In [15]:
query = '''
    SELECT name, population
    FROM countries
    ORDER BY population DESC
    LIMIT 10;
'''
pd.read_sql_query(query, conn_countries)

name  population
0          China  1330044000
1          India  1173108018
2  United States   310232863
3      Indonesia   242968342
4         Brazil   201103330
5       Pakistan   184404791
6     Bangladesh   156118464
7        Nigeria   154000000
8         Russia   140702000
9          Japan   127288000

### Exercise 1.e

Execute a query to select the name and area of the three countries with the largest area that is smaller than one million square km, ordered in descending order by area. Assign the query to the variable `query_limit` and the result to the variable `result_limit`.

In [16]:
#// BEGIN_TODO [ORG_1e] Define and execute a query (1 point)

query_limit = '''
    SELECT name, area
    FROM countries
    WHERE area < 1e6
    ORDER BY area DESC
    LIMIT 3
'''
result_limit = pd.read_sql_query(query_limit, conn_countries)
#// END_TODO [ORG_1e]

result_limit

name      area
0   Tanzania  945087.0
1    Nigeria  923768.0
2  Venezuela  912050.0

### Joining Tables

A benefit of a relational database is that data from different tables can be related together. For example, when selecting countries from the `'countries'` table we may like to return the full name of their continent, instead of the continent code. We can achieve this by joining the data from the `'countries'` and `'continents'` columns. First, let's see what the `'continents'` table contains:

In [17]:
pd.read_sql_query('SELECT * FROM continents;', conn_countries)

code           name
0   AF         Africa
1   AN     Antarctica
2   AS           Asia
3   EU         Europe
4   NM  North America
5   OC        Oceania
6   SM  South America

Note that the column `'continent'` in `countries` uses the exact same encoding for continents as the column '`code`' in `continents`. It serves as the **primary key** in `continents`.

Now, let's **join** data from the two tables `countries` and `continents` in such a way that each row of the top 10 countries sorted in descending order by population size will also show the full name of the continent of that country:

In [18]:
query = '''
    SELECT countries.name, population, continents.name AS 'continent name'
    FROM countries, continents
    WHERE countries.continent = continents.code
    ORDER BY population DESC
    LIMIT 10;
'''
pd.read_sql_query(query, conn_countries)

name  population continent name
0          China  1330044000           Asia
1          India  1173108018           Asia
2  United States   310232863  North America
3      Indonesia   242968342           Asia
4         Brazil   201103330  South America
5       Pakistan   184404791           Asia
6     Bangladesh   156118464           Asia
7        Nigeria   154000000         Africa
8         Russia   140702000         Europe
9          Japan   127288000           Asia

> **Notes:**
> * The column name `name` appears in both tables `countries` and `continents`. Therefore, we need to prefix it with the table name to disambiguate it.
> * In the result, we want the column with continent names to be labeled `'continent name'`. Therefore, we use the _alias_ `AS 'continent name'`. The quotes are required because of the whitespace character in the alias.

### Exercise 1.f

Execute a query to retrieve the country with the largest population on the continent with the name `'Europe'`. The result should contain two columns: `'country'` and `'population'`, containing the name of the country and its population.

Assign the query to the variable `query_join_population` and the result to the variable `result_join_population`.

In [19]:
#// BEGIN_TODO [ORG_1f] Max population in Europe (1 point)

query_join_population = '''
    SELECT countries.name AS 'country', MAX(population)
    FROM countries, continents
    WHERE continents.name = 'Europe' AND countries.continent = continents.code
    ORDER BY population DESC
'''

result_join_population = pd.read_sql_query(query_join_population, conn_countries)
#// END_TODO [ORG_1f]

result_join_population

country  MAX(population)
0  Russia        140702000

### Exercise 1.g

Execute a query to retrieve the names of countries on the continent with name `'Europe'` that have an area smaller than 100 square km. The result should contain two columns: `'country'` and `'area'`, containing the name of the country and its area.

Assign the query to the variable `query_join_area` and the result to the variable `result_join_area`.

In [20]:
#// BEGIN_TODO [ORG_1g] Small European countries (1 point)

query_join_area = '''
    SELECT countries.name AS 'country', area
    FROM countries, continents
    WHERE continents.name = 'Europe' AND countries.continent = continents.code AND area < 100
'''

result_join_area = pd.read_sql_query(query_join_area, conn_countries)
#// END_TODO [ORG_1g]

result_join_area

country   area
0   Gibraltar   6.50
1    Guernsey  78.00
2      Monaco   1.95
3  San Marino  61.20
4     Vatican   0.44

### Exercise 1.h

Execute a query to compute the total population per continent. The result should contain two columns: `'continent name'` (containing full names of the continents) and `'total population'`.

Assign the query to the variable `query_join` and the result to the variable `result_join`.

In [21]:
#// BEGIN_TODO [ORG_1h] Population per continent (1 point)

query_join = '''
    SELECT continents.name AS 'continent name', SUM(population) AS 'total population'
    FROM countries, continents
    WHERE continents.code = countries.continent
    GROUP BY continents.name
'''

result_join = pd.read_sql_query(query_join, conn_countries)
#// END_TODO [ORG_1h]

result_join

continent name  total population
0         Africa        1016879730
1     Antarctica               170
2           Asia        4130586341
3         Europe         739141025
4  North America         540105043
5        Oceania          36066049
6  South America         400143568

### Subqueries or Nested Queries

The result of one `SELECT` query can be combined with another `SELECT` query.

For instance, `SELECT` queries can be combined by 

* **`UNION`** to obtain the union of the `SELECT` query results
* **`INTERSECT`** to obtain the intersection of the `SELECT` query results
* **`EXCEPT`** to obtain the rows in the first result without those occurring in the second result

These correspond to the **union**, **intersection** and **relative complement** operations on sets. See the lecture slides and <a href="https://en.wikipedia.org/wiki/Set_(mathematics)#Basic_operations">Set § Basic operations</a> for more details.  
In some cases, the same result can be obtained by using **`AND`**, **`OR`**, and **`AND NOT`**
in the `WHERE` clause.

We give one example using `EXCEPT` that cannot simply be replicated using `AND NOT`.

* First, we define and execute a query to select (distinct) continents
    that have countries whose name starts with an `'E'`.
    
> We use the `LIKE` operator in the `WHERE` clause with the wild card `%`, which matches zero, one, or multiple characters. In this case, `name LIKE 'E%'` matches countries with a name that starts with an `'E'` and is followed by zero, one, or multiple characters.

* Second, we define and execute a query to select (distinct) continents
    that have countries whose area exceeds ten million square km.
* Third, we define and execute a query to select continents
    that have countries whose name starts with an `'E'`
    but that do _not_ have countries whose area exceeds ten million square km.

In [22]:
query_E = '''
SELECT DISTINCT continent
FROM countries
WHERE name LIKE 'E%'
'''
pd.read_sql_query(query_E, conn_countries)

continent
0        SM
1        EU
2        AF
3        NM
4        OC

In [23]:
query_large = '''
SELECT DISTINCT continent
FROM countries
WHERE area > 10000000
'''
pd.read_sql_query(query_large, conn_countries)

continent
0        AN
1        EU

In [24]:
query_E_except_large = query_E + 'EXCEPT' + query_large
print(query_E_except_large)
pd.read_sql_query(query_E_except_large, conn_countries)


SELECT DISTINCT continent
FROM countries
WHERE name LIKE 'E%'
EXCEPT
SELECT DISTINCT continent
FROM countries
WHERE area > 10000000



continent
0        AF
1        NM
2        OC
3        SM

> **Note:** Query `query_E` does not end with a semicolon. If it ended with a semicolon, then `query_E_except_large` would not be a valid SQL query.

### Exercise 1.i

Execute a query to select those continents that

* have countries whose name ends in `'d'`, and
* do _not_ have countries whose population exceeds one hundred million.

Assign the query to the variable `query_nested` and the result to the variable `result_nested`.

In [25]:
#// BEGIN_TODO [ORG_1i] Find the continents (1 point)

query_nested = '''
    SELECT DISTINCT continent
    FROM countries
    WHERE name LIKE '%d'
    EXCEPT
    SELECT DISTINCT continent
    FROM countries
    WHERE population > 100e6
'''
result_nested = pd.read_sql_query(query_nested, conn_countries)
#// END_TODO [ORG_1i]

result_nested

continent
0        AN
1        OC

## Step 3. Close Connection to Database

Once you no longer want to work with an open database, you can (and are encouraged to) close the connection to it:

In [26]:
conn_countries.close()

## 2. Data Cleaning

**Raw data** is often not in a form where it can easily be processed.
There can be all kinds of smaller or bigger issues with data:

* extra whitespace at the start and/or end of a string;
* missing values;
* duplicates;
* inconsistencies.

The process of preparing data for further processing is called **data cleaning**. In this section, we deal with a few common issues, which we will address with Pandas functions for data frames and series.

For the examples, we use the data in the `'datasets/countries.csv'` file.

In [27]:
country_data = pd.read_csv('datasets/countries.csv', na_values=[''], keep_default_na=False)
country_data.head()

name alpha_3  tld continent           capital      area  \
0               Andorra     AND  .ad        EU  Andorra la Vella     468.0   
1  United Arab Emirates     ARE  .ae        AS         Abu Dhabi   82880.0   
2           Afghanistan     AFG  .af        AS             Kabul  647500.0   
3   Antigua and Barbuda     ATG  .ag        NM        St. John's     443.0   
4              Anguilla     AIA  .ai        NM        The Valley     102.0   

   population  
0       84000  
1     4975593  
2    29121286  
3       86754  
4       13254

> **Note:** `df` denotes a `DataFrame` object, and `ts` a `Series` object. As usual, `pd` stands for the Pandas library itself.

### `df.astype()`: Convert to Given Type

With **`df.astype()`** you convert a data frame to a given type.
The `.dtypes` attribute of a data frame contains the column types:

In [28]:
country_data.dtypes

name           object
alpha_3        object
tld            object
continent      object
capital        object
area          float64
population      int64
dtype: object

We can see that columns `'name'`, ... , `'capital'` are of type `object` (which is also used for columns containing strings), `'area'` contains floating point numbers and `'population'` contains integers.

> **Remark:** The suffix `64` in both `float64` and `int64` denotes the number of bits used to store values of that type. A higher number of bits means a higher precision. Both of these types come from the `numpy` library. The built-in `float` and `int` types have different precision (see [here](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex)). However, they can still be used in `astype` to convert column types.

We can convert all column types in the data frame to strings as follows:

In [29]:
country_data_as_str = country_data.astype(str)
country_data_as_str.dtypes

name          object
alpha_3       object
tld           object
continent     object
capital       object
area          object
population    object
dtype: object

We can also convert the type of particular columns by supplying a dictionary as the argument of `astype()`:

In [30]:
population_as_str = country_data.astype({'population': str})
population_as_str.dtypes

name           object
alpha_3        object
tld            object
continent      object
capital        object
area          float64
population     object
dtype: object

### Exercise 2.a

Change the type of the `'area'` column in `country_data` to `int` and assign the resulting data frame to `area_as_int`.
There should be only one assignment in your code (namely for `area_as_int`).

In [31]:
#// BEGIN_TODO [ORG_2a] Area as int (1 point)

area_as_int = country_data.astype({'area' : int})
#// END_TODO [ORG_2a]

area_as_int.dtypes

name          object
alpha_3       object
tld           object
continent     object
capital       object
area           int32
population     int64
dtype: object

### `pd.to_numeric()`: Convert to Numeric

With **`pd.to_numeric()`** you can convert an argument to a numeric type. A suitable numeric type (e.g., `int` or `float`) will be chosen.
The argument can be a _list_ or a _Series_.

In [32]:
pd.to_numeric(country_data_as_str['population']).head()

0       84000
1     4975593
2    29121286
3       86754
4       13254
Name: population, dtype: int64

In [33]:
pd.to_numeric(country_data_as_str['area']).head()

0       468.0
1     82880.0
2    647500.0
3       443.0
4       102.0
Name: area, dtype: float64

### `ts.unique()`: Find All Unique Values

With `ts.unique()` you obtain an array in which each item of the series appears once.
The order in the array is in the order of appearance in the series object.

> **Note:** This returns an array, instead of a series.

In [34]:
country_data['capital'].unique()

array(['Andorra la Vella', 'Abu Dhabi', 'Kabul', "St. John's",
       'The Valley', 'Tirana', 'Yerevan', 'Luanda', nan, 'Buenos Aires',
       'Pago Pago', 'Vienna', 'Canberra', 'Oranjestad', 'Mariehamn',
       'Baku', 'Sarajevo', 'Bridgetown', 'Dhaka', 'Brussels',
       'Ouagadougou', 'Sofia', 'Manama', 'Bujumbura', 'Porto-Novo',
       'Gustavia', 'Hamilton', 'Bandar Seri Begawan', 'Sucre', 'Brasilia',
       'Nassau', 'Thimphu', 'Gaborone', 'Minsk', 'Belmopan', 'Ottawa',
       'West Island', 'Kinshasa', 'Bangui', 'Brazzaville', 'Bern',
       'Yamoussoukro', 'Avarua', 'Santiago', 'Yaounde', 'Beijing',
       'Bogota', 'San Jose', 'Havana', 'Praia', ' Willemstad',
       'Flying Fish Cove', 'Nicosia', 'Prague', 'Berlin', 'Djibouti',
       'Copenhagen', 'Roseau', 'Santo Domingo', 'Algiers', 'Quito',
       'Tallinn', 'Cairo', 'El-Aaiun', 'Asmara', 'Madrid', 'Addis Ababa',
       'Helsinki', 'Suva', 'Stanley', 'Palikir', 'Torshavn', 'Paris',
       'Libreville', 'London', "St. Geor

### Exercise 2.b

Find all unique continent names in the `country_data` data frame and assign the result to `unique_continents`.
There should be only one assignment in your code (namely for `unique_continents`).

In [35]:
#// BEGIN_TODO [ORG_2b] Unique continents (1 point)

unique_continents = country_data['continent'].unique()
#// END_TODO [ORG_2b]

unique_continents

array(['EU', 'AS', 'NM', 'AF', 'AN', 'SM', 'OC'], dtype=object)

### `df.duplicated()`: Find Duplicates

With **`df.duplicated()`** you obtain a _boolean mask_ indicating which rows appear more than once.
With the argument **`subset=`** you can restrict the search for duplicates to specific columns (either a single column or a list of columns).

In [36]:
bm_dup_capital = country_data.duplicated(subset='capital')
country_data[bm_dup_capital]

name alpha_3  tld continent   capital  \
29     Bonaire, Saint Eustatius and Saba      BES  .bq        NM       NaN   
33                          Bouvet Island     BVT  .bv        AN       NaN   
95      Heard Island and McDonald Islands     HMD  .hm        AN       NaN   
163                        Norfolk Island     NFK  .nf        OC  Kingston   
220                               Tokelau     TKL  .tk        OC       NaN   
232  United States Minor Outlying Islands     UMI  .um        OC       NaN   

      area  population  
29   328.0       18012  
33    49.0           0  
95   412.0           0  
163   34.6        1828  
220   10.0        1466  
232    0.0           0

There is an extra argument `keep`:

* **`keep='first'`** (default): Mark duplicates as `True` _except for the first occurrence_.
* **`keep='last'`** : Mark duplicates as `True` _except for the last occurrence_.
* **`keep=False`**`: Mark all duplicates as `True`.

In [37]:
bm_dup_capital_all = country_data.duplicated(subset='capital', keep=False)
country_data[bm_dup_capital_all]

name alpha_3  tld continent   capital  \
8                              Antarctica     ATA  .aq        AN       NaN   
29     Bonaire, Saint Eustatius and Saba      BES  .bq        NM       NaN   
33                          Bouvet Island     BVT  .bv        AN       NaN   
95      Heard Island and McDonald Islands     HMD  .hm        AN       NaN   
111                               Jamaica     JAM  .jm        NM  Kingston   
163                        Norfolk Island     NFK  .nf        OC  Kingston   
220                               Tokelau     TKL  .tk        OC       NaN   
232  United States Minor Outlying Islands     UMI  .um        OC       NaN   

           area  population  
8    14000000.0           0  
29        328.0       18012  
33         49.0           0  
95        412.0           0  
111     10991.0     2847232  
163        34.6        1828  
220        10.0        1466  
232         0.0           0

### Exercise 2.c

Some countries in `country_data` share the same `'tld'`. Find the list of `'tld'` values that are shared by more than one country and assign it to `duplicate_tld` (it should be a numpy array or a `list` and contain each shared `'tld'` only once).
There should be only one assignment in your code (namely for `duplicate_tld`).

In [38]:
#// BEGIN_TODO [ORG_2c] Duplicated data (1 point)

duplicate_tld = country_data['tld'][country_data.duplicated('tld')].unique().tolist()
#// END_TODO [ORG_2c]

duplicate_tld

['.gp', nan]

### `df.drop_duplicates()`: Remove Duplicates

With **`df.drop_duplicates()`** you can remove duplicated rows.
Using the argument **`subset=`**, you can restrict the search for duplicates to specific columns.
Below we do so for the `'capital'` column.

In [39]:
country_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251 entries, 0 to 250
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        251 non-null    object 
 1   alpha_3     251 non-null    object 
 2   tld         249 non-null    object 
 3   continent   251 non-null    object 
 4   capital     245 non-null    object 
 5   area        251 non-null    float64
 6   population  251 non-null    int64  
dtypes: float64(1), int64(1), object(5)
memory usage: 13.9+ KB


In [40]:
country_data.drop_duplicates(subset=['capital']).info()

<class 'pandas.core.frame.DataFrame'>
Index: 245 entries, 0 to 250
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        245 non-null    object 
 1   alpha_3     245 non-null    object 
 2   tld         243 non-null    object 
 3   continent   245 non-null    object 
 4   capital     244 non-null    object 
 5   area        245 non-null    float64
 6   population  245 non-null    int64  
dtypes: float64(1), int64(1), object(5)
memory usage: 15.3+ KB


Just as for `df.duplicated()`, there is an extra keyword argument `keep`:

* **`keep='first'`** (default): Drop duplicates _except for the first occurrence_.
* **`keep='last'`**: Drop duplicates _except for the last occurrence_.
* **`keep=False`**: Drop all duplicates.

### Exercise 2.d

Drop from the `country_data` data frame **all** countries which share the same `'tld'` and `'continent'` with another country.
Assign the resulting data frame to `dropped_dup_tld`.
There should be only one assignment in your code (namely for `dropped_dup_tld`).

In [41]:
#// BEGIN_TODO [ORG_2d] Drop duplicated data (1 point)

dropped_dup_tld = country_data.drop_duplicates(subset=['tld', 'continent'], keep=False)
#// END_TODO [ORG_2d]

dropped_dup_tld.info()

<class 'pandas.core.frame.DataFrame'>
Index: 248 entries, 0 to 250
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        248 non-null    object 
 1   alpha_3     248 non-null    object 
 2   tld         246 non-null    object 
 3   continent   248 non-null    object 
 4   capital     242 non-null    object 
 5   area        248 non-null    float64
 6   population  248 non-null    int64  
dtypes: float64(1), int64(1), object(5)
memory usage: 15.5+ KB


### `df.dropna()`: Remove Rows with Missing Values

In the EDA programming exercises the `.isnull()` method was introduced to select rows with missing values (_Not-Available_ (NA) and _Not-a-Number_ (NaN)). With **`df.dropna()`** you can remove rows with *missing values* in a data frame or series object.

With the argument **`how=`** you can indicate whether rows with at least one missing value or with all values missing are dropped:

* **`how='any'`** (default): if any value in a row is missing, drop that row
* **`how='all'`**: if all values in a row are missing, drop that row

In [42]:
country_data_dropna = country_data.dropna()
country_data_dropna.info()

<class 'pandas.core.frame.DataFrame'>
Index: 243 entries, 0 to 250
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        243 non-null    object 
 1   alpha_3     243 non-null    object 
 2   tld         243 non-null    object 
 3   continent   243 non-null    object 
 4   capital     243 non-null    object 
 5   area        243 non-null    float64
 6   population  243 non-null    int64  
dtypes: float64(1), int64(1), object(5)
memory usage: 15.2+ KB


In the output of the previous cell, we see that `country_data_dropna` has 243 entries (rows), and each of the columns has 243 non-null values.
This indicates that all missing values were removed.
With `.isnull()` we can check explicitly that the rows with missing values in the `'capital'` columns were indeed removed:

In [43]:
country_data[country_data['capital'].isnull()]

name alpha_3  tld continent capital  \
8                              Antarctica     ATA  .aq        AN     NaN   
29     Bonaire, Saint Eustatius and Saba      BES  .bq        NM     NaN   
33                          Bouvet Island     BVT  .bv        AN     NaN   
95      Heard Island and McDonald Islands     HMD  .hm        AN     NaN   
220                               Tokelau     TKL  .tk        OC     NaN   
232  United States Minor Outlying Islands     UMI  .um        OC     NaN   

           area  population  
8    14000000.0           0  
29        328.0       18012  
33         49.0           0  
95        412.0           0  
220        10.0        1466  
232         0.0           0

In [44]:
country_data_dropna[country_data_dropna['capital'].isnull()]

Empty DataFrame
Columns: [name, alpha_3, tld, continent, capital, area, population]
Index: []

### Exercise 2.e

Drop only the rows in `country_data` which have missing values in the `'tld'` column. Assign the resulting data frame to `dropped_na_tld`.
There should be only one assignment in your code (namely for `dropped_na_tld`).

> **Hint:** Look up the documentation of `dropna()` (e.g., via Shift-Tab-Tab).

In [45]:
#// BEGIN_TODO [ORG_2e] Drop missing data (1 point)

dropped_na_tld = country_data.dropna(subset='tld')
#// END_TODO [ORG_2e]

dropped_na_tld.info()

<class 'pandas.core.frame.DataFrame'>
Index: 249 entries, 0 to 250
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        249 non-null    object 
 1   alpha_3     249 non-null    object 
 2   tld         249 non-null    object 
 3   continent   249 non-null    object 
 4   capital     243 non-null    object 
 5   area        249 non-null    float64
 6   population  249 non-null    int64  
dtypes: float64(1), int64(1), object(5)
memory usage: 15.6+ KB


### `df.fillna()`: Replace Missing Values

With **`df.fillna()`** or **`ts.fillna()`**
you can replace *missing values*
in a data frame or series object.  
As always, note that this returns a new copy with the filled object.

In [46]:
country_data_filled = country_data['capital'].fillna(value='Unknown')
country_data_filled[country_data_filled.duplicated(keep=False)]

8       Unknown
29      Unknown
33      Unknown
95      Unknown
111    Kingston
163    Kingston
220     Unknown
232     Unknown
Name: capital, dtype: object

You can replace missing values in particular columns by providing a dictionary for the first argument, specifying which value to use for which column.

In [47]:
country_data_filled = country_data.fillna({'capital': 'Unknown'})
country_data_filled[country_data_filled.duplicated(subset='capital', keep=False)]

name alpha_3  tld continent   capital  \
8                              Antarctica     ATA  .aq        AN   Unknown   
29     Bonaire, Saint Eustatius and Saba      BES  .bq        NM   Unknown   
33                          Bouvet Island     BVT  .bv        AN   Unknown   
95      Heard Island and McDonald Islands     HMD  .hm        AN   Unknown   
111                               Jamaica     JAM  .jm        NM  Kingston   
163                        Norfolk Island     NFK  .nf        OC  Kingston   
220                               Tokelau     TKL  .tk        OC   Unknown   
232  United States Minor Outlying Islands     UMI  .um        OC   Unknown   

           area  population  
8    14000000.0           0  
29        328.0       18012  
33         49.0           0  
95        412.0           0  
111     10991.0     2847232  
163        34.6        1828  
220        10.0        1466  
232         0.0           0

### Exercise 2.f

Fill the missing values in the `'tld'` column of the `country_data` data frame with `'Missing'`.
Assign the resulting data frame to `filled_tld`.
There should be only one assignment in your code (namely for `filled_tld`).

In [48]:
#// BEGIN_TODO [ORG_2f] Fill missing data (1 point)

filled_tld = country_data[['tld']].fillna('Missing')
#// END_TODO [ORG_2f]

filled_tld.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251 entries, 0 to 250
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   tld     251 non-null    object
dtypes: object(1)
memory usage: 2.1+ KB


### `ts.str.strip()`: Strip Characters from Begin/End of Strings

With **`ts.str.strip()`** you can remove characters from the beginning and end of all strings in a `Series` object.  
Without argument, it strips all leading and trailing **whitespace**.

Here we strip all trailing and leading zeros from the area that was converted to a string
in an earlier example.

In [49]:
country_data_as_str['area'].head()

0       468.0
1     82880.0
2    647500.0
3       443.0
4       102.0
Name: area, dtype: object

In [50]:
country_data_as_str['area'].str.strip('0').head()

0       468.
1     82880.
2    647500.
3       443.
4       102.
Name: area, dtype: object

> **Remark:** If a string of more than one character is passed to `ts.str.strip()`all such characters will be stripped regardless of their order, which might not be what you expect (see the example below). To strip specific prefixes or suffixes, use `str.removeprefix()` or `str.removesuffix()`. For other string methods, see [here](https://docs.python.org/3/library/stdtypes.html#string-methods).

In [51]:
country_data_as_str['area'].str.strip('These letters are irrelevant. Only this 0, the two periods and this 8 matter in this case.').head()

0      46
1       2
2    6475
3     443
4     102
Name: area, dtype: object

### Exercise 2.g

Remove the character `'s'` from the beginning and end of all the country names in `country_data`.
Assign the resulting series to `stripped_names`.
There should be only one assignment in your code (namely for `stripped_names`).

In [52]:
#// BEGIN_TODO [ORG_2g] Strip the 's' (1 point)

stripped_names = country_data['name'].str.removeprefix('s').str.rstrip('s')
#// END_TODO [ORG_2g]

stripped_names.head()

0                Andorra
1    United Arab Emirate
2            Afghanistan
3    Antigua and Barbuda
4               Anguilla
Name: name, dtype: object

## 3. Exercises: Cleaning the Mouse Experiment


In the earlier exercises you were closely guided. In the following exercises we will put your knowledge to the test.

Note that these exercises may seem more difficult as you will need to work more indepedently. When you struggle with an exercise then go back to the corresponding earlier section and make sure you really understand the introduced concepts. Do not hesitate to experiment with your own code!


In the following exercises we will analyze the data from the mouse experiment.

### Exercise 3.a

Read the data from `'datasets/paths_props.csv'` into a data frame and assign it `df_props`.

In [53]:
#// BEGIN_TODO [ORG_3a] Read the data (1 point)

df_props = pd.read_csv('datasets/paths_props.csv')

#// END_TODO [ORG_3a]

df_props.head()

user  trial  use_tue_laptop  mouse_speed  mouse_accuracy  right_handed  \
0  1226      5               1          1.0             1.0             1   
1  1226      6               1          1.0             1.0             1   
2  1226      7               1          1.0             1.0             1   
3  1226      8               1          1.0             1.0             1   
4  1226      9               1          1.0             1.0             1   

   input_method                   major  target_x  target_y  target_radius  \
0             0  Industrial Engineering       -84       -84              6   
1             0  Industrial Engineering         0       190              6   
2             0  Industrial Engineering      -158      -158              6   
3             0  Industrial Engineering        81       -81              3   
4             0  Industrial Engineering      -157       157              9   

      delay  
0  2.136719  
1  3.540913  
2  3.688381  
3  2.823134  
4  3.690253

Each row describes a single trial, i.e., one movement from the center to the target. The meaning of the columns is the following:

- `'user'`: the id of the user that recorded this trial;
- `'trial'`: the trial number. Trial numbers are local to each user;
- `'use_tue_laptop'`: 1 means a TU/e laptop was used during this trial, 0 means a different laptop was used;
- `'mouse_speed'`, `'mouse_accuracy'`: 1 means the user confirmed this mouse setting, 0 or `NaN` means the user did not confirm this mouse setting;  
    For example, users who claim to use a non-TU/e laptop are not shown the `'mouse_speed'` and `'mouse_accuracy'` options, so the user cannot confirm those.
- `'right_handed'`: 1 means user writes with their right hand, 0 means user writes with their left hand;
- `'input_method'`: 0 means a trackpad, 1 means a mouse;
- `'target_x'`, `'target_y'`: the screen coordinates in pixels of the blue target, relative to the red dot in the center;
- `'target_radius'`: radius of the blue target in pixels;
- `'delay'`: delay in seconds between the start of the trial (i.e., the moment that the mouse was placed on the red dot) and the moment that the blue target was shown;
- `'major'`: the major of the user.

As you can see, there is a lot of redundancy in the data. We will reorganize the data into two tables: one describing the user settings (which do not change), and one describing the individual trials. These two tables will be linked via the `'user'` column.

### Exercise 3.b

Inspect the `df_props` data frame. You will notice that the `'mouse_speed'` and `'mouse_accuracy'` columns contain NaN  values. They are the result of a user using a non-TU/e laptop and not having to confirm these mouse settings. Replace the NaN values in `'mouse_speed'` and `'mouse_accuracy'` columns by 0.

In [54]:
#// BEGIN_TODO [ORG_3b] Fill missing data (1 point)

df_props = df_props.fillna({'mouse_speed' : 0, 'mouse_accuracy' : 0})
#// END_TODO [ORG_3b]

df_props.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33342 entries, 0 to 33341
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   user            33342 non-null  int64  
 1   trial           33342 non-null  int64  
 2   use_tue_laptop  33342 non-null  int64  
 3   mouse_speed     33342 non-null  float64
 4   mouse_accuracy  33342 non-null  float64
 5   right_handed    33342 non-null  int64  
 6   input_method    33342 non-null  int64  
 7   major           33342 non-null  object 
 8   target_x        33342 non-null  int64  
 9   target_y        33342 non-null  int64  
 10  target_radius   33342 non-null  int64  
 11  delay           33342 non-null  float64
dtypes: float64(3), int64(8), object(1)
memory usage: 3.1+ MB


### Exercise 3.c

Convert the `'mouse_speed'` and `'mouse_accuracy'` columns to `int` type.

In [55]:
#// BEGIN_TODO [ORG_3c] Convert to integers (1 point)

df_props = df_props.astype({'mouse_speed':int, 'mouse_accuracy':int})

#// END_TODO [ORG_3c]

df_props.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33342 entries, 0 to 33341
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   user            33342 non-null  int64  
 1   trial           33342 non-null  int64  
 2   use_tue_laptop  33342 non-null  int64  
 3   mouse_speed     33342 non-null  int32  
 4   mouse_accuracy  33342 non-null  int32  
 5   right_handed    33342 non-null  int64  
 6   input_method    33342 non-null  int64  
 7   major           33342 non-null  object 
 8   target_x        33342 non-null  int64  
 9   target_y        33342 non-null  int64  
 10  target_radius   33342 non-null  int64  
 11  delay           33342 non-null  float64
dtypes: float64(1), int32(2), int64(8), object(1)
memory usage: 2.8+ MB


### Exercise 3.d

Create a data frame `df_trials` from `df_props` with only the columns `'user'`, `'trial'`, `'input_method'`, `'target_x'`, `'target_y'`, `'target_radius'`, `'delay'`.

In [56]:
#// BEGIN_TODO [ORG_3d] Collect trials information (1 point)

df_trials = df_props[['user', 'trial', 'input_method', 'target_x', 'target_y', 'target_radius', 'delay']]

#// END_TODO [ORG_3d]

df_trials.head()

user  trial  input_method  target_x  target_y  target_radius     delay
0  1226      5             0       -84       -84              6  2.136719
1  1226      6             0         0       190              6  3.540913
2  1226      7             0      -158      -158              6  3.688381
3  1226      8             0        81       -81              3  2.823134
4  1226      9             0      -157       157              9  3.690253

### Exercise 3.e

Create a data frame `df_users` from `df_props` with only the columns `'user'`, `'use_tue_laptop'`, `'mouse_speed'`, `'mouse_accuracy'`, `'right_handed'`, `'major'`, **avoiding** duplicate rows.

In [57]:
#// BEGIN_TODO [ORG_3e] Collect users information (1 point)

df_users = df_props[['user', 'use_tue_laptop', 'mouse_speed', 'mouse_accuracy', 'right_handed', 'major']].drop_duplicates()

#// END_TODO [ORG_3e]

df_users.head()

user  use_tue_laptop  mouse_speed  mouse_accuracy  right_handed  \
0    1226               1            1               1             1   
30   1937               1            1               1             1   
60    625               1            1               1             0   
90   1025               1            1               1             0   
120   620               1            1               1             1   

                      major  
0    Industrial Engineering  
30   Industrial Engineering  
60          Applied Physics  
90      Innovation Sciences  
120       Industrial Design

### Exercise 3.f

Inspect the `df_users` data frame. You will notice that for some users there is more than one row. For example, some users appear to have indicated multiple different computer settings. Since at this point we cannot ask those users which settings they actually used, we will discard those users and the corresponding trials.

Identify which users have entered contradicting information, i.e., both right- and left-handed or using different computer settings, and assign them to a list `wrong_users` (it should be a numpy array or a `list` and contain each user with contradicting information only once).

> **Hint:** For these users there will be multiple rows in the data frame.

In [58]:
#// BEGIN_TODO [ORG_3f] Find the wrong users (1 point)

df_users

user  use_tue_laptop  mouse_speed  mouse_accuracy  right_handed  \
0      1226               1            1               1             1   
30     1937               1            1               1             1   
60      625               1            1               1             0   
90     1025               1            1               1             0   
120     620               1            1               1             1   
...     ...             ...          ...             ...           ...   
33207  1578               0            0               0             1   
33237  2349               1            1               1             1   
33267   447               1            1               1             1   
33297  2088               0            0               0             0   
33312  1131               1            1               1             1   

                        major  
0      Industrial Engineering  
30     Industrial Engineering  
60            Applied Physics  
90        Innovation Sciences  
120         Industrial Design  
...                       ...  
33207       Industrial Design  
33237       Industrial Design  
33267       Industrial Design  
33297       Industrial Design  
33312       Built Environment  

[1184 rows x 6 columns]

In [59]:
wrong_users = df_users[df_users.duplicated('user', keep=False)]['user'].unique().tolist()

In [60]:
#// END_TODO [ORG_3f]

wrong_users

[1598, 1017]

### Exercise 3.g

Discard the rows from `df_users` and `df_trials` corresponding to the users which have entered contradicting information.

> **Hint:** You could use a `for` loop to iterate through users with contradicting information. For each user remove the rows corresponding to that user.

In [61]:
#// BEGIN_TODO [ORG_3g] Discard the wrong users (1 point)

for user in df_users[df_users.duplicated(subset=['user'], keep=False)]['user']:
    # Remove rows from df_users
    df_users = df_users[df_users['user'] != user]
for user in df_trials[df_trials.duplicated(subset=['user'], keep=False)]['user']:
    # Remove rows from df_trials
    df_trials = df_trials[df_trials['user'] != user]
#// END_TODO [ORG_3g]

## 4. Exercises: Analyzing the Mouse Experiment

In these exercises we will analyze the mouse experiment data to see if the experiment was properly conducted. We will use the data stored in the SQLite database in the `'datasets/mouse_experiment.db'` file. This database contains two tables, analogous to the two data frames you created in the previous exercises: 

- the **users** table, which contains columns `'user'`, `'use_tue_laptop'`, `'mouse_speed'`, `'mouse_accuracy'`, `'right_handed'`, `'major'`
- the **trials** table, which contains columns `'user'`, `'trial'`, `'input_method'`, `'target_x'`, `'target_y'`, `'target_radius'`, `'delay'`

These two tables are linked via the `'user'` column. The meaning of the other columns is in the `'datasets/paths_props.csv'` file described above. 

### Exercise 4.a

Connect to the SQLite database `'datasets/mouse_experiment.db'`, assigning the connection to the variable **`conn_experiment`**.

In [62]:
#// BEGIN_TODO [ORG_4a] Connect to mouse experiment database (1 point)

conn_experiment = sqlite3.connect('datasets/mouse_experiment.db')

#// END_TODO [ORG_4a]

### Exercise 4.b

Execute an SQL query to count the number of trials that were performed by user 2 with a mouse.

Assign the query to the variable `query_user_mouse`, and the result (an integer number) to the variable `num_user_mouse`.

In [63]:
#// BEGIN_TODO [ORG_4b] Count the trials (1 point)

query_user_mouse = '''
    SELECT count(trial)
    FROM trials
    WHERE user == 2
'''

num_user_mouse = pd.read_sql_query(query_user_mouse, conn_experiment)
#// END_TODO [ORG_4b]

num_user_mouse

count(trial)
0            30

### Exercise 4.c

Execute a query to count the number of trials that were performed with a trackpad.

Assign the query to the variable `query_all_trackpad`, and the result (an integer number) to the variable `num_all_trackpad`.

In [64]:
#// BEGIN_TODO [ORG_4c] Count the trackpad trials (1 point)

query_all_trackpad = '''
    SELECT count(trial)
    FROM trials
    WHERE input_method == 0
'''
num_all_trackpad = pd.read_sql_query(query_all_trackpad, conn_experiment)
#// END_TODO [ORG_4c]

num_all_trackpad

count(trial)
0         16792

### Exercise 4.d

Execute a query to count the number of trials that were performed by left-handed people.

Assign the query to the variable `query_lefthanded`, and the result (an integer number) to the variable `num_lefthanded`.

> **Hint:** You will need to combine data from both `trials` and `users` tables.

In [65]:
#// BEGIN_TODO [ORG_4d] Count the lefthanded trials (1 point)

query_lefthanded = '''
    SELECT COUNT(trials.trial) AS num_lefthanded
    FROM trials
    JOIN users ON trials.user = users.user
    WHERE users.right_handed = 0
'''



num_lefthanded = pd.read_sql_query(query_lefthanded, conn_experiment)

#// END_TODO [ORG_4d]

num_lefthanded

num_lefthanded
0            4316

### Exercise 4.e

Execute a query to compute the average delay between the start of a trial and the moment the target was shown for left- and right-handed people separately.

Assign the query to the variable `query_delay`, and the result: a data frame with two rows, first row for left-handed and second row for right-handed people, and one column `'average delay'`, to the variable `result_delay`.

> **Hint:** Use the `GROUP BY` clause.

In [66]:
#// BEGIN_TODO [ORG_4e] Average delay (1 point)

query_delay = '''
    SELECT AVG(delay) AS 'average delay'
    FROM users, trials
    WHERE users.user == trials.user
    GROUP BY right_handed
'''

result_delay = pd.read_sql_query(query_delay, conn_experiment)
#// END_TODO [ORG_4e]

result_delay

average delay
0       2.979530
1       2.996293

### Exercise 4.f

Let us check whether the experiment was properly conducted. In particular, let us check whether the users working with a mouse were not disadvantaged compared to those using the trackpad. We will do this in two steps. First, execute a query to compute the average radius of targets that were shown to users using the different input methods.

Assign the query to the variable `query_radius`, and the result (a data frame with two columns `'method'` and `'average radius'`) to the variable `result_radius`.

> **Hint:** Use the `GROUP BY` clause to aggregate the different methods.

In [67]:
#// BEGIN_TODO [ORG_4f] Average radius (1 point)

query_radius = '''
    SELECT input_method AS 'method', AVG(target_radius) AS 'average radius'
    FROM users, trials
    WHERE users.user = trials.user
    GROUP BY input_method
'''
result_radius = pd.read_sql_query(query_radius, conn_experiment)

#// END_TODO [ORG_4f]

result_radius

method  average radius
0       0        6.015900
1       1        6.001091

### Exercise 4.g

We consider the experiment to be properly conducted if the absolute difference in the average target radius between the two input methods is at most 20% of the smaller value. Was the experiment properly conducted?

Assign your boolean answer to `proper_experiment`.

> **Remark:** The margin of 20% is somewhat arbitrary here. In the exercise set about hypothesis testing a more rigorous way of testing whether the averages are not too far apart will be shown.

In [68]:
#// BEGIN_TODO [ORG_4g] Proper experiment? (1 point)

proper_experiment = np.abs(result_radius.iloc[0, 1] / result_radius.iloc[1, 1]) - 1 <= 0.2

#// END_TODO [ORG_4g]

proper_experiment

True

### Exercise 4.h

Close the connection to the `'mouse_experiment.db'` database.

In [69]:
#// BEGIN_TODO [ORG_4h] Close the mouse experiment database (1 point)

conn_experiment.close()
#// END_TODO [ORG_4h]

# Feedback

Please fill in this questionaire to help us improve this course for the next year. Your feedback will be anonymized and will not affect your grade in any way!

### How many hours did you spend on these Exercises?

Assign a number to `feedback_time`.

In [70]:
#// BEGIN_FEEDBACK [Feedback_1] (0 point)
feedback_time = 6
#// END_FEEDBACK [Feedback_1] (0 point)

import numbers
assert isinstance(feedback_time, numbers.Number), "Please assign a number to feedback_time"
feedback_time

6

### How difficult did you find these Exercises?

Assign an integer to `feedback_difficulty`, on a scale 0 - 10, with 0 being very easy, 5 being just right, and 10 being very difficult.

In [71]:
#// BEGIN_FEEDBACK [Feedback_2] (0 point)
feedback_difficulty = 6
#// END_FEEDBACK [Feedback_2] (0 point)

import numbers
assert isinstance(feedback_difficulty, numbers.Number), "Please assign a number to feedback_difficulty"
feedback_difficulty

6

### (Optional) What did you like?

Assign a string to `feedback_like`.

In [72]:
#// BEGIN_FEEDBACK [Feedback_3] (0 point)

#// END_FEEDBACK [Feedback_3] (0 point)

### (Optional) What can be improved?

Assign a string to `feedback_improve`. Please be specific, so that we can act on your feedback. For example, mention the specific exercises and what was unclear.

In [73]:
#// BEGIN_FEEDBACK [Feedback_4] (0 point)

#// END_FEEDBACK [Feedback_4] (0 point)




## How to Submit Your Work

1. **Before submitting**, you must run your notebook by doing **Kernel > Restart & Run All**.  
   Make sure that your notebook runs without errors **in linear order**.
1. Remember to rename the notebook, replacing `...-template.ipynb` with `...-yourIDnr.ipynb`, where `yourIDnr` is your TU/e identification number.
1. Submit the executed notebook with your work
   for the appropriate assignment in **Canvas**.
1. In the **Momotor** tab in Canvas,
  you can select that assignment again to find some feedback on your submitted work.
  If there are any problems reported by _Momotor_,
  then you need to fix those,
  and **resubmit the fixed notebook**.

In case of a high workload on our server
(because many students submit close to the deadline),
it may take longer to receive the feedback.




---

In [74]:
# List all defined names
%whos

Variable                 Type          Data/Info
------------------------------------------------
AUTHOR_DATE              str           2023-12-07
AUTHOR_ID_NR             str           1920847
AUTHOR_NAME              str           Shiyun Kong
HTML                     type          <class 'IPython.core.display.HTML'>
area_as_int              DataFrame                          name<...>n\n[251 rows x 7 columns]
bm_dup_capital           Series        0      False\n1      Fals<...>nLength: 251, dtype: bool
bm_dup_capital_all       Series        0      False\n1      Fals<...>nLength: 251, dtype: bool
conn_countries           Connection    <sqlite3.Connection object at 0x0000020F25D54940>
conn_experiment          Connection    <sqlite3.Connection object at 0x0000020F25D55740>
country_data             DataFrame                          name<...>n\n[251 rows x 7 columns]
country_data_as_str      DataFrame                          name<...>n\n[251 rows x 7 columns]
country_data_dropna      D

---

# (End of Notebook) <span class="tocSkip"></span>

&copy; 2017-2023 - **TU/e** - Eindhoven University of Technology